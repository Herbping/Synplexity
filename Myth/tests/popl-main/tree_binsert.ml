type cmp =
  | CEq
  | CGt
  | CLt

type nat =
  | Z
  | S of nat

type tree =
  | L
  | N of tree * nat * tree

let rec comp_nat (n1:nat) (n2:nat) : cmp =
  match n1 with
  | Z u -> (match n2 with
          | Z u -> CEq
          | S n2 -> CLt)
  | S n1 -> (match n2 with
              | Z u -> CGt
              | S n2 -> comp_nat n1 n2)
;;

let rec tree_binsert : tree -> nat -> tree |>
/\(L -> /\( 0 -> N (L, 0, L)
          , 1 -> N (L, 1, L)
          , 2 -> N (L, 2, L))
  , N (L, 1, L) -> /\( 0 -> N (N (L, 0, L), 1, L)
                     , 1 -> N (L, 1, L)
                     , 2 -> N (L, 1, N (L, 2, L)))
  , N (L, 0, L) -> /\( 0 -> N (L, 0, L)
                     , 1 -> N (L, 0, N (L, 1, L))
                     , 2 -> N (L, 0, N (L, 2, L)))
  , N (L, 2, L) -> /\( 0 -> N (N (L, 0, L), 2, L)
                     , 1 -> N (N (L, 1, L), 2, L)
                     , 2 -> N (L, 2, L))
  , N (N (L, 0, L), 1, L) ->
    /\( 0 -> N (N (L, 0, L), 1, L)
      , 1 -> N (N (L, 0, L), 1, L)
      , 2 -> N (N (L, 0, L), 1, N(L, 2, L)))
  , N (L, 0, N (L, 1, L)) ->
      /\( 2 -> N (L, 0, N (L, 1, N(L, 2, L))))
  , N (N (L, 1, L), 2, L) ->
      /\( 0 -> N (N (N(L, 0, L), 1, L), 2, L))
  , N (L, 1, N (L, 2, L)) ->
      /\( 0 -> N (N (L, 0, L), 1, N (L, 2, L))
        , 1 -> N (L, 1, N (L, 2, L)))
  , N (N (L, 1, L), 2, L) ->
      /\( 0 -> N (N (N(L, 0, L), 1, L), 2, L))
  ) = ?
