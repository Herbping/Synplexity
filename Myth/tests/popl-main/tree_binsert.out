(* Generated Thu Oct  8 20:41:13 2015 *)
fix f1 (t1:tree) : nat -> tree ->
  fix f2 (n1:nat) : tree ->
    match t1 with
    | L u1 -> N (L, n1, L)
    | N t2 -> match (comp_nat n1) (#2 t2) with
      | CEq u1 -> N (#1 t2, n1, #3 t2)
      | CGt u1 -> N (#1 t2, #2 t2, (f1 (#3 t2)) n1)
      | CLt u1 -> N ((f1 (#1 t2)) n1, #2 t2, #3 t2)
